### PR TITLE
Probe ipc4 update

### PR DESCRIPTION
**Pull request by jsarha**
Drop IPC3 probe DMA framing and use IPC4 framing for IPC3 mode too. This change simplifies the code quite a bit. After this change both the SOF firmware and sof-probe demux tool should be updated at the same time.

Following this change there is several fixes and cleanups.

### CODE REVIEW

**Comment by jsarha** [comment]
@abonislawski could you review these changes?

File: tools/probes/probes_main.c [hunk 1]
```diff
@@ -209,17 +209,6 @@ int process_sync(struct probe_data_packet **packet, uint8_t **w_ptr, uint32_t *t
 	return 0;
 }
 
-static bool sync_word_at(uint8_t *buf, size_t len)
-{
-	if (len < sizeof(uint32_t))
```
Comment (lgirdwood) [inline tools/probes/probes_main.c:223]: Can you say in the commit message why the condition can be simplified (i.e. why do we not need length anymore)

File: tools/probes/probes_main.c [hunk 2]
```diff
@@ -267,45 +267,34 @@ void parse_data(char *file_in)
 	memset(&data, 0, DATA_READ_LIMIT);
 	memset(&files, 0, sizeof(struct wave_files) * FILES_LIMIT);
 
+	start = 0;
 	/* data read loop to process DATA_READ_LIMIT bytes at each iteration */
```
Comment (kv2019i) [inline tools/probes/probes_main.c:244]: Is the comment still valid when start is added?

Comment (jsarha) [reply tools/probes/probes_main.c]: Pretty much, if not counting the minor exception of moving 1-3 bytes remainder to the beginning of buffer  if looking for the next sync word... Anyway updated the comment.

File: src/include/ipc/probe.h [hunk 3]
```diff
@@ -3,6 +3,8 @@
  * Copyright(c) 2022 Intel Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (jsarha) [reply src/include/ipc/probe.h:3]: Thanks. will fix.

File: tools/probes/probes_main.c (hunk 1 above)
Comment (jsarha) [reply tools/probes/probes_main.c:223]: Sure. Updated the commit message.

File: tools/probes/probes_main.c [hunk 4]
```diff
@@ -182,39 +183,44 @@ void finalize_wave_files(struct wave_files *files)
...
 int process_sync(struct probe_data_packet *packet, uint8_t **w_ptr, uint32_t *total_data_to_copy)
 {
 	struct probe_data_packet *temp_packet;
 
-	/* request to copy data_size from probe packet */
-	*total_data_to_copy = packet->data_size_bytes;
+	/* request to copy data_size from probe packet and 64-bit checksum */
+	*total_data_to_copy = packet->data_size_bytes + sizeof(uint64_t);
 
 	if (packet->data_size_bytes > PACKET_MAX_SIZE) {
```
Comment (jsarha) [inline tools/probes/probes_main.c]: Oops... this is a bug, should be (*total_data_to_copy > PACKET_MAX_SIZE).

Comment (jsarha) [reply tools/probes/probes_main.c]: Actually even the original was buggy.... the condition should be:
(sizeof(struct probe_data_packet) + *total_data_to_copy > PACKET_MAX_SIZE)

And this still not that great. Its reallocing the packet time and time again, every time packet is bigger than PACKET_MAX_SIZE. But this should be harmless waste of cycles.

File: tools/probes/probes_main.c [hunk 5]
```diff
@@ -182,39 +183,44 @@ void finalize_wave_files(struct wave_files *files)
...
+			  packet->buffer_id  +
+			  packet->format +
+			  packet->timestamp_high +
+			  packet->timestamp_low +
+			  packet->data_size_bytes);
+
+	checksump = (uint64_t *) (packet->data + packet->data_size_bytes);
+
+	if (sum != *checksump) {
+		fprintf(stderr, "Checksum error 0x%016lx != 0x%016lx\n", sum, *checksump);
```
Comment (lyakh) [inline tools/probes/probes_main.c:184]: is `lx` enough to print 64 bits? Don't you need `llx`?

Comment (jsarha) [reply tools/probes/probes_main.c:184]: Apparently it is. I got a warning when I first tried llx. I think both long long int and long int are 64-bit in 64-bit x86 architecture, and it depends on how the int64_t is defined whether llx or lx should de used. Pretty annoying actually.

Comment (paulstelian97) [reply tools/probes/probes_main.c:184]: https://en.cppreference.com/w/cpp/header/cinttypes There's the <inttypes.h> header that provides some nice printf format string macros that work well when you're not trying to do any padding or other custom formatting. See if they are useful. (for example, PRIx64).

**Comment by plbossart** [comment]
@jsarha for the probe definitions, we need the FLEX_ARRAY thing that is required for upstream, see the topology example.

````diff
diff --git a/include/uapi/sound/asoc.h b/include/uapi/sound/asoc.h
index 053949287ce8..dd8ad790fc15 100644
--- a/include/uapi/sound/asoc.h
+++ b/include/uapi/sound/asoc.h
@@ -226,9 +226,9 @@ struct snd_soc_tplg_vendor_array {
  	__le32 type;	/* SND_SOC_TPLG_TUPLE_TYPE_ */
  	__le32 num_elems;	/* number of elements in array */
  	union {
-		struct snd_soc_tplg_vendor_uuid_elem uuid[0];
-		struct snd_soc_tplg_vendor_value_elem value[0];
-		struct snd_soc_tplg_vendor_string_elem string[0];
+		__DECLARE_FLEX_ARRAY(struct snd_soc_tplg_vendor_uuid_elem, uuid);
+		__DECLARE_FLEX_ARRAY(struct snd_soc_tplg_vendor_value_elem, value);
+		__DECLARE_FLEX_ARRAY(struct snd_soc_tplg_vendor_string_elem, string);
  	};
  } __attribute__((packed));
  
````

**Comment by jsarha** [comment]
> @jsarha for the probe definitions, we need the FLEX_ARRAY thing that is required for upstream, see the topology example.

So this was actually meant for: https://github.com/thesofproject/linux/pull/3818 , no?

Should I add FLEX_ARRAY change to the beginning of the series there?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
> > @jsarha for the probe definitions, we need the FLEX_ARRAY thing that is required for upstream, see the topology example.
> 
> So this was actually meant for: [thesofproject/linux#3818](https://github.com/thesofproject/linux/pull/3818) , no?
> 
> Should I add FLEX_ARRAY change to the beginning of the series there?

Yes, needed for kernel.  Probably a nice thing to have in Zephyr/SOF eventually.

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX can you check CI, the IPC4 logs are timing out. Thanks !

File: tools/probes/probes_main.c (hunk 5 above)
Comment (jsarha) [reply tools/probes/probes_main.c:184]: The macro appear to be work with padding too. Thanks, will update.

**Pull request merged**